File: android/gemcore/src/main/kotlin/com/gemwallet/android/model/PaymentAsset.kt
package com.gemwallet.android.model

import com.gemwallet.android.domains.asset.chain
import com.gemwallet.android.ext.isValidAddress
import com.wallet.core.primitives.PaymentRequest

internal sealed interface PaymentAsset {
    data object Unsupported : PaymentAsset

    data class Single(val assetInfo: AssetInfo) : PaymentAsset

    data class Choice(val assets: List<AssetInfo>) : PaymentAsset

    companion object {
        fun from(request: PaymentRequest, assets: List<AssetInfo>): PaymentAsset {
            val payable = when (val assetId = request.assetId) {
                null -> assets.filter { it.asset.chain.isValidAddress(request.address) }
                else -> assets.filter { it.asset.id == assetId }
            }
            return when (payable.size) {
                0 -> Unsupported
                1 -> Single(payable.first())
                else -> Choice(payable)
            }
        }
    }
}

File: ios/Packages/PrimitivesComponents/Sources/Types/PaymentAsset.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import GemstonePrimitives
import Primitives

enum PaymentAsset {
    case unsupported
    case single(AssetData)
    case choice([AssetData])
}

extension PaymentAsset {
    static func from(_ payment: PaymentRequest, assets: [AssetData]) -> PaymentAsset {
        let payable = payableAssets(for: payment, in: assets)

        guard let assetData = payable.first else { return .unsupported }
        return payable.count == 1 ? .single(assetData) : .choice(payable)
    }

    private static func payableAssets(for payment: PaymentRequest, in assets: [AssetData]) -> [AssetData] {
        guard let assetId = payment.assetId else {
            return assets.filter { $0.asset.chain.isValidAddress(payment.address) }
        }
        return assets.filter { $0.asset.id == assetId }
    }
}
